pkg/controller/reconciler/deploymentresource_reconciler.go

```diff
@@ -0,0 +1,374 @@
+
+	deploymentResource.Status.Operation = nil
+	deploymentResource.Status.Phrase = radappiov1alpha3.DeploymentResourcePhraseFailed
+	err := r.Client.Status().Update(ctx, deploymentResource)
+	if err != nil {
+		return ctrl.Result{}, err
+	}
+
+	return ctrl.Result{}, nil
+}
+
+func (r *DeploymentResourceReconciler) reconcileDelete(ctx context.Context, deploymentResource *radappiov1alpha3.DeploymentResource) (ctrl.Result, error) {
+	logger := ucplog.FromContextOrDiscard(ctx)
+
+	logger.Info("Resource is being deleted.", "resourceId", deploymentResource.Spec.Id)
+
+	// Since we're going to reconcile, update the observed generation.
+	//
+	// We don't want to do this if we're in the middle of an operation, because we haven't
+	// fully processed any status changes until the async operation completes.
+	deploymentResource.Status.ObservedGeneration = deploymentResource.Generation
+
+	// NOTE: The following is a workaround for Radius API behavior. Since deleting
+	// an application or environment can leave hanging resources, we need to make sure to
+	// delete these resources before deleting the application or environment.
+
+	// Check other resources that depend on this resource.
+	// List all DeploymentResource objects in the same namespace
+	// that have the same rootFileName.
+	deploymentResourceList := &radappiov1alpha3.DeploymentResourceList{}
+	err := r.Client.List(ctx, deploymentResourceList, client.InNamespace(deploymentResource.Namespace), client.MatchingFields{rootFileNameField: deploymentResource.Spec.RootFileName})
+	if err != nil {
+		return ctrl.Result{}, nil
+	}
+
+	appsCount := 0
+	envsCount := 0
+	otherCount := 0
+	for _, dr := range deploymentResourceList.Items {
+		if dr.Status.Phrase == radappiov1alpha3.DeploymentResourcePhraseDeleted {
+			continue
+		}
+		if strings.Contains(dr.Spec.Id, "Applications.Core/applications") {
+			appsCount++
+		} else if strings.Contains(dr.Spec.Id, "Applications.Core/environments") {
+			envsCount++
+		} else if dr.Spec.Id != "" {
+			logger.Info("Resource is being used by another resource.", "resourceId", dr.Spec.Id)
+			otherCount++
+		}
+	}
+
+	if strings.Contains(deploymentResource.Spec.Id, "Applications.Core/applications") {
+		// dont delete app until otherCount is 0
+		if otherCount > 0 {
+			logger.Info("Resource is an application, being used by another resource.", "resourceId", deploymentResource.Spec.Id)
+			return ctrl.Result{Requeue: true, RequeueAfter: r.requeueDelay()}, nil
+		}
+	}
+
+	if strings.Contains(deploymentResource.Spec.Id, "Applications.Core/environments") {
+		if otherCount > 0 {
+			logger.Info("Resource is an environment, being used by another resource.", "resourceId", deploymentResource.Spec.Id)
+			return ctrl.Result{Requeue: true, RequeueAfter: r.requeueDelay()}, nil
+		}
+	}
+
+	poller, err := r.startDeleteOperation(ctx, deploymentResource)
```

Comment:
It looks like this code us directly manipulating the state of the DR resource, and handling its lifecycle. We should talk about this, I was expecting something different. 

What I had in mind for when a delete happens on a DT:

- DT controller **initiates** the delete of the DR through the K8s API and uses a **watch** to observe completion.
- DR controller is responsible for the deletion of underlying resource **owned** by the DR.